src/tool_application_support.cpp

```diff
@@ -0,0 +1,117 @@
+#include <clangmetatool/tool_application_support.h>
+
+#include <clang/Basic/Diagnostic.h>
+#include <clang/Driver/Compilation.h>
+#include <clang/Driver/Driver.h>
+#include <clang/Driver/ToolChain.h>
+#include <llvm/ADT/SmallVector.h>
+#include <llvm/Option/Option.h>
+#include <llvm/Support/ErrorHandling.h>
+#include <llvm/Support/FileSystem.h>
+#include <llvm/Support/Host.h>
+#include <llvm/Support/Path.h>
+
+#include <llvm/Config/llvm-config.h>
+#if LLVM_VERSION_MAJOR >= 8
+#include <llvm/Support/VirtualFileSystem.h>
+#else
+#include <clang/Basic/VirtualFileSystem.h>
+#endif
+
+#include <cstring>
+#include <memory>
+#include <sstream>
+
+namespace clangmetatool {
+
+void ToolApplicationSupport::verifyInstallation(
+    const clang::tooling::CompilationDatabase &compilations,
+    const std::vector<std::string> &sourcePathList) {
+  clang::DiagnosticsEngine diagnostics(new clang::DiagnosticIDs,
+                                       new clang::DiagnosticOptions,
+                                       new clang::IgnoringDiagConsumer);
+
+  // Check each compile command, each may have different settings for toolchain
+  // and resource directory.
+
+  for (const auto &source : sourcePathList) {
+    for (const auto &command : compilations.getCompileCommands(source)) {
+      std::vector<std::string> args = command.CommandLine;
+
+      // Create a driver and toolchain to verify
+
+      clang::driver::Driver driver(args[0], llvm::sys::getDefaultTargetTriple(),
+                                   diagnostics);
+
+      std::vector<const char *> argsArray;
+      for (const auto &arg : args) {
+        argsArray.push_back(arg.c_str());
+      }
+
+      std::unique_ptr<clang::driver::Compilation> compilation(
+          driver.BuildCompilation(llvm::makeArrayRef(argsArray)));
+
+      const clang::driver::ToolChain &toolChain =
+          compilation->getDefaultToolChain();
+
+      // Check that we can find C++ includes using the toolchain
+
+      llvm::opt::InputArgList inputArgs;
+      llvm::opt::ArgStringList cxxIncludes;
+      toolChain.AddClangCXXStdlibIncludeArgs(inputArgs, cxxIncludes);
+
+      bool found = false;
+      for (const auto &include : cxxIncludes) {
+        // Skip options
+
+        if (include[0] == '-') {
+          continue;
+        }
+
+        // Check if a sample C++ header can be found at this path
+
+        llvm::SmallVector<char, 128> path(include,
+                                          include + std::strlen(include));
+        llvm::sys::path::append(path, "iostream");
```

Comment:
I don't see a check that this is a c++ translation unit, if you're invoking as the C compiler, does it make sense to find for the c++ header?

Reply:
Interesting, I'll add a test. I don't think it matters at this point in the initialization, it finds the same toolchain regardless of whether you are compiling for c or c++.

Reply:
I tested with a c file, looks like the checks still pass and fail the same way. I don't think there is a c mode or anything, that is just for individual files. You can even pass c and c++ files at the same time.